### ISSUE TITLE
[S3] Observed a panic: "invalid memory address or nil pointer dereference"

### ISSUE THREAD

**Issue by dervos** [aws-controllers-k8s/community#955]
**Describe the bug**
After I drained a node the ack-s3-controller started on a different node, after this the panic occurs. Related: #918

**Steps to reproduce**

**Expected outcome**
* The controller shouldn't crash
* It should print out a readable error message

**Environment**

* Kubernetes version: 1.21
* Using EKS (yes/no), if so version? EKS 1.21
* AWS service targeted: S3 v0.0.3

```
2021-09-14T09:30:47.754Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": "0.0.0.0:8080"}
2021-09-14T09:30:47.799Z	INFO	setup	initializing service controller	{"aws.service": "s3"}
2021-09-14T09:30:47.800Z	INFO	setup	starting manager	{"aws.service": "s3"}
2021-09-14T09:30:47.800Z	INFO	controller-runtime.manager	starting metrics server	{"path": "/metrics"}
2021-09-14T09:30:47.800Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "adoptedresource", "source": "kind source: /, Kind="}
2021-09-14T09:30:47.800Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "bucket", "source": "kind source: /, Kind="}
2021-09-14T09:30:47.901Z	INFO	controller-runtime.controller	Starting Controller	{"controller": "adoptedresource"}
2021-09-14T09:30:47.901Z	INFO	controller-runtime.controller	Starting Controller	{"controller": "bucket"}
2021-09-14T09:30:47.901Z	INFO	controller-runtime.controller	Starting workers	{"controller": "adoptedresource", "worker count": 1}
2021-09-14T09:30:47.901Z	INFO	controller-runtime.controller	Starting workers	{"controller": "bucket", "worker count": 1}
E0914 09:30:48.715742       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 361 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1680ce0, 0x27cdf70)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1680ce0, 0x27cdf70)
	/usr/local/go/src/runtime/panic.go:967 +0x166
github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket.customPreCompare(0xc0000a2c88, 0xc0000a2100)
	/github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket/hook.go:303 +0x95d
github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket.newResourceDelta(0xc0000a2c88, 0xc0000a2100, 0x0)
	/github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket/delta.go:41 +0x75
github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket.(*resourceDescriptor).Delta(0x2816810, 0x1c3a380, 0xc0000a2c88, 0x1c3a380, 0xc0000a2100, 0x0)
	/github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket/descriptor.go:70 +0x54
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).updateResource(0xc000188000, 0x1c24b80, 0xc000576ea0, 0x1c2e320, 0xc000125100, 0x1c3a380, 0xc0000a2c88, 0x1c3a380, 0xc0000a2100, 0x0, ...)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:323 +0x1d5
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).Sync(0xc000188000, 0x1c24b80, 0xc000576ea0, 0x1c2e320, 0xc000125100, 0x1c3a380, 0xc0000a2c88, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:221 +0x78d
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).reconcile(0xc000188000, 0x1c24b80, 0xc000576ea0, 0x1c2e320, 0xc000125100, 0x1c3a380, 0xc0000a2c88, 0xc, 0x7ffd80c80b8c, 0x0, ...)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:181 +0x130
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).Reconcile(0xc000188000, 0xc0005115d0, 0x7, 0xc00051bc60, 0x1b, 0x0, 0xc04838d1f5b5bdd2, 0xc00001b950, 0xc00001b8c8)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:165 +0x7e0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00018e000, 0x16e2560, 0xc0005788a0, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:256 +0x161
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00018e000, 0x203000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232 +0xae
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc00018e000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00056e5b0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00056e5b0, 0x1beb9a0, 0xc000576ae0, 0x1, 0xc0000fc4e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00056e5b0, 0x3b9aca00, 0x0, 0xc00032ae01, 0xc0000fc4e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc00056e5b0, 0x3b9aca00, 0xc0000fc4e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:193 +0x305
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x151501d]

goroutine 361 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1680ce0, 0x27cdf70)
	/usr/local/go/src/runtime/panic.go:967 +0x166
github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket.customPreCompare(0xc0000a2c88, 0xc0000a2100)
	/github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket/hook.go:303 +0x95d
github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket.newResourceDelta(0xc0000a2c88, 0xc0000a2100, 0x0)
	/github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket/delta.go:41 +0x75
github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket.(*resourceDescriptor).Delta(0x2816810, 0x1c3a380, 0xc0000a2c88, 0x1c3a380, 0xc0000a2100, 0x0)
	/github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket/descriptor.go:70 +0x54
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).updateResource(0xc000188000, 0x1c24b80, 0xc000576ea0, 0x1c2e320, 0xc000125100, 0x1c3a380, 0xc0000a2c88, 0x1c3a380, 0xc0000a2100, 0x0, ...)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:323 +0x1d5
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).Sync(0xc000188000, 0x1c24b80, 0xc000576ea0, 0x1c2e320, 0xc000125100, 0x1c3a380, 0xc0000a2c88, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:221 +0x78d
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).reconcile(0xc000188000, 0x1c24b80, 0xc000576ea0, 0x1c2e320, 0xc000125100, 0x1c3a380, 0xc0000a2c88, 0xc, 0x7ffd80c80b8c, 0x0, ...)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:181 +0x130
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).Reconcile(0xc000188000, 0xc0005115d0, 0x7, 0xc00051bc60, 0x1b, 0x0, 0xc04838d1f5b5bdd2, 0xc00001b950, 0xc00001b8c8)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:165 +0x7e0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00018e000, 0x16e2560, 0xc0005788a0, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:256 +0x161
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00018e000, 0x203000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232 +0xae
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc00018e000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00056e5b0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00056e5b0, 0x1beb9a0, 0xc000576ae0, 0x1, 0xc0000fc4e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00056e5b0, 0x3b9aca00, 0x0, 0xc00032ae01, 0xc0000fc4e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc00056e5b0, 0x3b9aca00, 0xc0000fc4e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:193 +0x305
```

**Comment by RedbackThomson** [comment]
This is the same issue as https://github.com/aws-controllers-k8s/community/issues/918 .

The nil dereference bug was fixed in https://github.com/aws-controllers-k8s/s3-controller/pull/41 , but we have not yet pushed a release which includes that new commit. If this is something you are blocked on, I am happy to cut a new one.

**Comment by stunthamster** [comment]
Hi @RedbackThomson if you could cut a new release that'd be super useful from my POV as I'm also seeing these issues a fair bit; would make things a bit easier vs maintaining my own version of the image and amending the helm chart.

**Comment by RedbackThomson** [comment]
Not a problem! Happy to do that now.

**Comment by stunthamster** [comment]
That's great, cheers!

Also, I _may_ have found another issue. I'm getting seg fault issuing a bucket confirmation. The details are below - happy to raise another issue if this is separate. Quickly glancing, it seems to have some issues with a CORS policy. It probably is invalid, but looks like its not being handled well. I've hit this one a fair bit with policies. 

Steps to reproduce
Create a bucket with the following config:
```

> apiVersion: s3.services.k8s.aws/v1alpha1
> kind: Bucket
> metadata:
>   name: tch-mdhh-primary
> spec:
>   name: tch-mdhh-primary
>   cors:
>     corsRules:
>     - allowedHeaders:
>       - ""
>     - allowedMethods:
>       - GET
>       - PUT
>       allowedOrigins:
>       - mydomainwouldgohere.com
>       exposeHeaders:
>       - Origin
>       - Content-Type
>       - Content-MD5
>       - Content-Disposition
>   policy: >
>     {
>       "Version": "2012-10-17",
>       "Id": "BlockAllObjects",
>       "Statement": [
>         {
>           "Effect": "Deny",
>           "Principal": "*",
>           "Action": "s3:*",
>           "Resource": "arn:aws:s3:::md-temp-test/*",
>           "Condition": {
>                 "Bool": {
>                     "aws:SecureTransport": "false"
>                 }
>             }
>         }
>       ]
>     }
> 
> 
```

Expected outcome
If policy is in error an error is raised and the controller does not crash/exit.

Environment

Kubernetes version: 1.21
Using EKS (yes/no), if so version? EKS 1.21
AWS service targeted: S3 v0.0.3

```
2021-09-15T20:58:00.126Z	ERROR	controller-runtime.controller	Reconciler error	{"controller": "bucket", "request": "default/tch-mdhh-primary", "error": "InvalidParameter: 2 validation error(s) found.\n- missing required field, PutBucketCorsInput.CORSConfiguration.CORSRules[0].AllowedMethods.\n- missing required field, PutBucketCorsInput.CORSConfiguration.CORSRules[0].AllowedOrigins.\n"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:258
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
E0915 20:58:01.322856       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 271 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1680ce0, 0x27cdf70)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1680ce0, 0x27cdf70)
	/usr/local/go/src/runtime/panic.go:967 +0x166
github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket.customPreCompare(0xc0006b2018, 0xc0000b62a0)
	/github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket/hook.go:303 +0x95d
github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket.newResourceDelta(0xc0006b2018, 0xc0000b62a0, 0x0)
	/github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket/delta.go:41 +0x75
github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket.(*resourceDescriptor).Delta(0x2816810, 0x1c3a380, 0xc0006b2018, 0x1c3a380, 0xc0000b62a0, 0x0)
	/github.com/aws-controllers-k8s/s3-controller/pkg/resource/bucket/descriptor.go:70 +0x54
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).updateResource(0xc0001c0d80, 0x1c24b80, 0xc00081e300, 0x1c2e320, 0xc0007e4200, 0x1c3a380, 0xc0006b2018, 0x1c3a380, 0xc0000b62a0, 0x0, ...)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:323 +0x1d5
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).Sync(0xc0001c0d80, 0x1c24b80, 0xc00081e300, 0x1c2e320, 0xc0007e4200, 0x1c3a380, 0xc0006b2018, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:221 +0x78d
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).reconcile(0xc0001c0d80, 0x1c24b80, 0xc00081e300, 0x1c2e320, 0xc0007e4200, 0x1c3a380, 0xc0006b2018, 0x9, 0x7ffea7aab9c1, 0x0, ...)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:181 +0x130
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).Reconcile(0xc0001c0d80, 0xc0008c7b30, 0x7, 0xc0008c7b20, 0x10, 0x0, 0xc048b5764795e905, 0xc0002e2cf0, 0xc0002e2c68)
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.12.0/pkg/runtime/reconciler.go:165 +0x7e0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001ceb40, 0x16e2560, 0xc00026bb80, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:256 +0x161
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001ceb40, 0xc000614600)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232 +0xae
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0001ceb40)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0007c2180)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0007c2180, 0x1beb9a0, 0xc0007af800, 0x1, 0xc00009c7e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0007c2180, 0x3b9aca00, 0x0, 0x1910b01, 0xc00009c7e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc0007c2180, 0x3b9aca00, 0xc00009c7e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:193 +0x305
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x151501d]

```

**Comment by RedbackThomson** [comment]
It appears you have an extra dash in your configuration:

```yaml
     corsRules:
     - allowedHeaders:
       - ""
     - allowedMethods:
       - GET
       - PUT
       allowedOrigins:
       - mydomainwouldgohere.com
       exposeHeaders:
       - Origin
       - Content-Type
       - Content-MD5
       - Content-Disposition
```
should be

```yaml
     corsRules:
     - allowedHeaders:
       - ""
       allowedMethods: <-- This line
       - GET
       - PUT
       allowedOrigins:
       - mydomainwouldgohere.com
       exposeHeaders:
       - Origin
       - Content-Type
       - Content-MD5
       - Content-Disposition
```

**Comment by stunthamster** [comment]
That'll be it! Again, thanks for the help.

**Comment by RedbackThomson** [comment]
@stunthamster S3 v0.0.4 released which contains the bug fix for the nil dereference panics - https://github.com/aws-controllers-k8s/s3-controller/releases/tag/v0.0.4

**Comment by stunthamster** [comment]
@RedbackThomson that release looks like it's resolved the issues I'm having, and formatting problems with policy etc look like they are being handled gracefully. Many thanks for the prompt release, its a big help!
